File: src/main/java/com/qrware/domain/inventory/MovementType.java
package com.qrware.domain.inventory;

public enum MovementType {
    
    RECEIPT("Receipt", "Item received into inventory"),
    ISSUE("Issue", "Item issued from inventory"),
    TRANSFER("Transfer", "Item transferred between locations"),
    MOVE("Move", "Item moved to different location"),
    ADJUSTMENT("Adjustment", "Inventory quantity adjustment"),
    CYCLE_COUNT("Cycle Count", "Cycle count adjustment"),
    PHYSICAL_COUNT("Physical Count", "Physical inventory count adjustment"),
    RESERVE("Reserve", "Item reserved for order"),
    UNRESERVE("Unreserve", "Item reservation removed"),
    PICK("Pick", "Item picked for order"),
    PACK("Pack", "Item packed for shipment"),
    SHIP("Ship", "Item shipped out"),
    RETURN("Return", "Item returned to inventory"),
    PUTAWAY("Putaway", "Item put away to storage location"),
    REPLENISHMENT("Replenishment", "Item replenished from bulk location"),
    ALLOCATION("Allocation", "Item allocated to specific purpose"),
    DEALLOCATION("Deallocation", "Item allocation removed"),
    QUARANTINE("Quarantine", "Item moved to quarantine"),
    RELEASE("Release", "Item released from quarantine"),
    HOLD("Hold", "Item put on hold"),
    UNHOLD("Unhold", "Item removed from hold"),
    DAMAGE("Damage", "Item marked as damaged"),
    DISPOSAL("Disposal", "Item disposed of"),
    LOSS("Loss", "Item marked as lost"),
    FOUND("Found", "Lost item found"),
    EXPIRY("Expiry", "Item marked as expired"),
    RECALL("Recall", "Item recalled"),
    STAGING("Staging", "Item moved to staging area"),
    CROSSDOCK("Crossdock", "Item moved for crossdocking"),
    CONSOLIDATION("Consolidation", "Items consolidated"),
    SPLIT("Split", "Item quantity split"),
    MERGE("Merge", "Items merged together"),
    CONVERSION("Conversion", "Unit conversion"),
    PRODUCTION("Production", "Item produced/manufactured"),
    CONSUMPTION("Consumption", "Item consumed in production"),
    SCRAP("Scrap", "Item scrapped"),
    REWORK("Rework", "Item sent for rework"),
    SAMPLE("Sample", "Item taken as sample"),
    LOAN("Loan", "Item loaned out"),
    LOAN_RETURN("Loan Return", "Loaned item returned"),
    
    ORDER_RECEIPT("Order Receipt", "Item received via order"),
    ORDER_ISSUE("Order Issue", "Item issued via order"),
    ORDER_PICK("Order Pick", "Item picked for order fulfillment"),
    ORDER_PACK("Order Pack", "Item packed for order shipment"),
    ORDER_CANCEL("Order Cancel", "Movement cancelled due to order cancellation"),
    ORDER_RETURN("Order Return", "Item returned from order"),
    ORDER_ADJUSTMENT("Order Adjustment", "Quantity adjusted due to order discrepancy");

    private final String displayName;
    private final String description;

    MovementType(String displayName, String description) {
        this.displayName = displayName;
        this.description = description;
    }

    public String getDisplayName() {
        return displayName;
    }

    public String getDescription() {
        return description;
    }

    public boolean isInbound() {
        return this == RECEIPT || this == RETURN || this == FOUND || 
               this == RELEASE || this == UNHOLD || this == PRODUCTION ||
               this == LOAN_RETURN || this == ORDER_RECEIPT || this == ORDER_RETURN;
    }

    public boolean isOutbound() {
        return this == ISSUE || this == SHIP || this == DISPOSAL || 
               this == LOSS || this == SCRAP || this == CONSUMPTION ||
               this == SAMPLE || this == LOAN || this == ORDER_ISSUE || 
               this == ORDER_PICK || this == ORDER_PACK;
    }

    public boolean isMovement() {
        return this == TRANSFER || this == MOVE || this == PUTAWAY || 
               this == REPLENISHMENT || this == STAGING || this == CROSSDOCK;
    }

    public boolean isStatusChange() {
        return this == RESERVE || this == UNRESERVE || this == QUARANTINE || 
               this == HOLD || this == DAMAGE || this == EXPIRY || this == RECALL;
    }

    public boolean isAdjustment() {
        return this == ADJUSTMENT || this == CYCLE_COUNT || this == PHYSICAL_COUNT;
    }

    public boolean isOrderRelated() {
        return this == RESERVE || this == PICK || this == PACK || this == SHIP ||
               this == ALLOCATION || this == DEALLOCATION || this == ORDER_RECEIPT ||
               this == ORDER_ISSUE || this == ORDER_PICK || this == ORDER_PACK ||
               this == ORDER_CANCEL || this == ORDER_RETURN || this == ORDER_ADJUSTMENT;
    }

    public boolean isQualityRelated() {
        return this == QUARANTINE || this == RELEASE || this == DAMAGE || 
               this == EXPIRY || this == RECALL || this == REWORK || this == SAMPLE;
    }

    public boolean isProductionRelated() {
        return this == PRODUCTION || this == CONSUMPTION || this == SCRAP || 
               this == REWORK || this == CONVERSION;
    }

    public boolean increasesQuantity() {
        return this == RECEIPT || this == RETURN || this == FOUND || 
               this == PRODUCTION || this == LOAN_RETURN || this == ORDER_RECEIPT || 
               this == ORDER_RETURN;
    }

    public boolean decreasesQuantity() {
        return this == ISSUE || this == SHIP || this == DISPOSAL || 
               this == LOSS || this == SCRAP || this == CONSUMPTION ||
               this == SAMPLE || this == LOAN || this == ORDER_ISSUE || 
               this == ORDER_PICK || this == ORDER_PACK;
    }

    public boolean requiresApproval() {
        return this == DISPOSAL || this == SCRAP || this == ADJUSTMENT ||
               this == LOSS || this == DAMAGE || this == ORDER_CANCEL ||
               this == ORDER_ADJUSTMENT;
    }

    public boolean isOrderSpecific() {
        return this == ORDER_RECEIPT || this == ORDER_ISSUE || this == ORDER_PICK ||
               this == ORDER_PACK || this == ORDER_CANCEL || this == ORDER_RETURN ||
               this == ORDER_ADJUSTMENT;
    }

    @Override
    public String toString() {
        return displayName;
    }
}

File: Wareapp/app/src/main/java/com/qrware/app/data/model/MovementType.kt
package com.qrware.app.data.model

enum class MovementType(
    val displayName: String,
    val description: String,
    val increasesQuantity: Boolean = false,
    val decreasesQuantity: Boolean = false,
    val requiresApproval: Boolean = false
) {
    RECEIPT("Przyjęcie", "Towar przyjęty do magazynu", increasesQuantity = true),
    ISSUE("Wydanie", "Towar wydany z magazynu", decreasesQuantity = true),
    TRANSFER("Transfer", "Towar przeniesiony między lokalizacjami"),
    MOVE("Przemieszczenie", "Towar przemieszczony do innej lokalizacji"),
    ADJUSTMENT("Korekta", "Korekta ilości towaru", requiresApproval = true),
    CYCLE_COUNT("Liczenie cykliczne", "Korekta z liczenia cyklicznego"),
    PHYSICAL_COUNT("Inwentaryzacja", "Korekta z inwentaryzacji fizycznej"),
    RESERVE("Rezerwacja", "Towar zarezerwowany pod zamówienie"),
    UNRESERVE("Zwolnienie rezerwacji", "Rezerwacja towaru zwolniona"),
    PICK("Kompletacja", "Towar skompletowany pod zamówienie"),
    PACK("Pakowanie", "Towar spakowany do wysyłki"),
    SHIP("Wysyłka", "Towar wysłany", decreasesQuantity = true),
    RETURN("Zwrot", "Towar zwrócony do magazynu", increasesQuantity = true),
    PUTAWAY("Odłożenie", "Towar odłożony na miejsce składowania"),
    REPLENISHMENT("Uzupełnienie", "Uzupełnienie towaru z lokalizacji masowej"),
    ALLOCATION("Alokacja", "Towar przydzielony do konkretnego celu"),
    DEALLOCATION("Dealokacja", "Alokacja towaru usunięta"),
    QUARANTINE("Kwarantanna", "Towar przeniesiony do kwarantanny"),
    RELEASE("Zwolnienie", "Towar zwolniony z kwarantanny"),
    HOLD("Wstrzymanie", "Towar wstrzymany"),
    UNHOLD("Zwolnienie wstrzymania", "Wstrzymanie towaru usunięte"),
    DAMAGE("Uszkodzenie", "Towar oznaczony jako uszkodzony", requiresApproval = true),
    DISPOSAL("Utylizacja", "Towar zutylizowany", decreasesQuantity = true, requiresApproval = true),
    LOSS("Utrata", "Towar oznaczony jako utracony", decreasesQuantity = true, requiresApproval = true),
    FOUND("Odnalezienie", "Utracony towar odnaleziony", increasesQuantity = true),
    EXPIRY("Wygaśnięcie", "Towar przeterminowany"),
    RECALL("Wycofanie", "Towar wycofany"),
    STAGING("Sortowanie", "Towar przeniesiony do strefy sortowania"),
    CROSSDOCK("Crossdocking", "Towar przekierowany przez crossdock"),
    CONSOLIDATION("Konsolidacja", "Towary skonsolidowane"),
    SPLIT("Podział", "Ilość towaru podzielona"),
    MERGE("Łączenie", "Towary połączone"),
    CONVERSION("Konwersja", "Konwersja jednostek"),
    PRODUCTION("Produkcja", "Towar wyprodukowany", increasesQuantity = true),
    CONSUMPTION("Zużycie", "Towar zużyty w produkcji", decreasesQuantity = true),
    SCRAP("Złomowanie", "Towar złomowany", decreasesQuantity = true, requiresApproval = true),
    REWORK("Przeróbka", "Towar wysłany do przeróbki"),
    SAMPLE("Próbka", "Pobranie próbki", decreasesQuantity = true),
    LOAN("Wypożyczenie", "Towar wypożyczony", decreasesQuantity = true),
    LOAN_RETURN("Zwrot wypożyczenia", "Wypożyczony towar zwrócony", increasesQuantity = true);

    fun isInbound(): Boolean = increasesQuantity

    fun isOutbound(): Boolean = decreasesQuantity

    fun isMovement(): Boolean = this in listOf(
        TRANSFER, MOVE, PUTAWAY, REPLENISHMENT, STAGING, CROSSDOCK
    )

    fun isStatusChange(): Boolean = this in listOf(
        RESERVE, UNRESERVE, QUARANTINE, HOLD, DAMAGE, EXPIRY, RECALL
    )

    fun isAdjustment(): Boolean = this in listOf(
        ADJUSTMENT, CYCLE_COUNT, PHYSICAL_COUNT
    )

    fun isOrderRelated(): Boolean = this in listOf(
        RESERVE, PICK, PACK, SHIP, ALLOCATION, DEALLOCATION
    )

    fun isQualityRelated(): Boolean = this in listOf(
        QUARANTINE, RELEASE, DAMAGE, EXPIRY, RECALL, REWORK, SAMPLE
    )

    fun isProductionRelated(): Boolean = this in listOf(
        PRODUCTION, CONSUMPTION, SCRAP, REWORK, CONVERSION
    )

    companion object {
        fun getInboundTypes(): List<MovementType> = values().filter { it.isInbound() }
        fun getOutboundTypes(): List<MovementType> = values().filter { it.isOutbound() }
        fun getMovementTypes(): List<MovementType> = values().filter { it.isMovement() }
        fun getAdjustmentTypes(): List<MovementType> = values().filter { it.isAdjustment() }
        fun getOrderRelatedTypes(): List<MovementType> = values().filter { it.isOrderRelated() }
        fun getQualityRelatedTypes(): List<MovementType> = values().filter { it.isQualityRelated() }
        fun getProductionRelatedTypes(): List<MovementType> = values().filter { it.isProductionRelated() }
        fun getApprovalRequiredTypes(): List<MovementType> = values().filter { it.requiresApproval }
    }
}